cmd/kube-controller-manager/app/core.go

```diff
@@ -65,6 +66,13 @@ import (
 	netutils "k8s.io/utils/net"
 )
 
+const (
+	// defaultNodeMaskCIDRIPv4 is default mask size for IPv4 node cidr
+	defaultNodeMaskCIDRIPv4 = 24
+	// defaultNodeMaskCIDRIPv6 is default mask size for IPv6 node cidr
+	defaultNodeMaskCIDRIPv6 = 98
```

Comment:
I was chatting with @aramase about this and he explained to me perfectly the consequences of this decision.
Setting 98 here impose a limitation in the Cluster CIDR because of this https://github.com/kubernetes/kubernetes/pull/79993/files#diff-a8ab00a26d321bed050bda3418c91d4dR46

This means that by default if we don't set the `node-cidr-mask-size`, the default cluster cidr minimum mask size will be 82 in IPv6.

I suggest setting the default mask length for an IPv6 cluster to /64, that's the minimum prefix recommended for end users in RFCs (i.e https://tools.ietf.org/html/rfc6177) and Internet registries (i.e. https://www.ripe.net/publications/docs/ripe-707) , that will imply `defaultNodeMaskCIDRIPv6 = 80`
@thockin @khenidak thoughts? does it make sense?

Reply:
I'm not an IPv6 expert so sorry in advance for naive questions. 

A default node cidr mask of `/80` is ~280 billion IP addresses, and with a cluster CIDR of `/64` that allows for ~65000 nodes, each with ~280 billion IP addresses. Even though both numbers seem pretty far from reach, to me it makes more sense to allow more nodes in the cluster CIDR and reduce # of IPs per node? A /98 would be 1 million addresses per node but now I can fit 17 million nodes in the `/64`. 

Pretty unrealistic for 1 cluster, but I'm thinking of clusters with pod native routing where pod IPs need to be unique across a single network. A user may pick 1 `/64` subnet for N (100s - 1000s) clusters. The /98 as a default node cidr mask would make much more sense in that case to me. 

> that will imply defaultNodeMaskCIDRIPv6 = 80

I saw in RFC 6177 the recommendation for /64s but I'm wondering how that implies /80 for the node mask cidr, can you clarify please?

Reply:
I'm not considering myself an expert either, just a person with some networking field experience ;)
When talking about ipv6 allocation you can't think in terms of ip addresses per networks, that's something we were used to do for ipv4 that no longer applies with ipv6. We should think in terms of [best practices](https://www.ripe.net/publications/docs/ripe-690), simplicity, route summarization, ... I recommend this article to go deeper in the topic https://etherealmind.com/allocating-64-wasteful-ipv6-not/

When talking about prefix assignment is important to take into account the following paragraph that you can find in the best practices I linked above

> To keep addressing plans usable and understandable, and to align with DNS reverse zone delegations, the size of the delegated prefix should align with a nibble boundary. Each hexadecimal character in an IPv6 prefix represents one nibble, which is 4 bits. The length of a delegated prefix should therefore always be a multiple of 4. 

Another limitation is the relation between the cluster and the the node cidr
> 	// The subnet mask size cannot be greater than 16 more than the cluster mask size
>	// TODO: https://github.com/kubernetes/kubernetes/issues/44918

In a purist approach, we should have a /64 per node and  a /48 per cluster ,( /52, /56, /60 per cluster cidr are valid and recommended ones too)

My suggestion was using /64 for the whole cluster, that because of the limitation, will imply 64+16 a /80 per node. But now that you asked the right question I'm thinking that maybe we should go with the /48 -> /64

Reply:
Thanks for the explanation! Looks like I have some reading to do :)

Reply:
Did some reading -- `/48` for cluster and `/64` for node as defaults seems reasonable for me.

Reply:
agreed! changed default IPv6 node cidr mask to `/64`.